{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import docx\n",
    "from docx import Document\n",
    "import pandas as pd\n",
    "import gsheet_to_case\n",
    "from docx.enum.section import WD_SECTION, WD_ORIENT\n",
    "from docx.oxml import OxmlElement\n",
    "from docx.oxml.ns import qn\n",
    "from docx.shared import Cm, Inches\n",
    "from docx.shared import Pt\n",
    "import numpy as np\n",
    "from datetime import date\n",
    "from docx.enum.text import WD_ALIGN_PARAGRAPH, WD_BREAK, WD_LINE_SPACING\n",
    "from datetime import datetime\n",
    "from docx.dml.color import ColorFormat\n",
    "from docx.shared import RGBColor\n",
    "\n",
    "pd.options.mode.chained_assignment = None  # default='warn'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get dataframe from google sheet\n",
    "df = gsheet_to_case.import_gsheet('GECO_DATABASE.xlsx','BATCH31_R&D')\n",
    "print(df.columns)\n",
    "df = df[['NO.','RITM Lab ID', 'UIC', 'AGE', 'SEX', 'PATIENT ADDRESS (CITY)', 'PATIENT ADDRESS (REGION)', 'DRU', 'DATE OF COLLECTION (MM-DD-YYYY)']]\n",
    "df.rename(columns = {'RITM Lab ID':'RITM Lab ID', 'PATIENT ADDRESS (CITY)': 'PATIENT ADDRESS','DRU':'SENDING FACILITY', 'DATE OF COLLECTION (MM-DD-YYYY)':'DATE OF COLLECTION'}, inplace = True)\n",
    "df['LINELIST REGION'] = ''\n",
    "df['OUTCOME'] = ''\n",
    "df['LINEAGE'] = ''\n",
    "df['REMARKS'] = ''\n",
    "df = df[['NO.','RITM Lab ID', 'UIC', 'AGE', 'SEX', 'PATIENT ADDRESS','LINELIST REGION', 'SENDING FACILITY', 'DATE OF COLLECTION', 'OUTCOME', 'LINEAGE', 'REMARKS' ]]\n",
    "df['DATE OF COLLECTION'] = df['DATE OF COLLECTION'].dt.strftime('%m/%d/%Y')\n",
    "df['AGE'] = df['AGE'].astype('Int64')\n",
    "df['NO.'] = df['NO.'].astype(np.int64)\n",
    "df = df.astype(str)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# open a document\n",
    "document = Document(\"linelist.docx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_paragraph(document, font_name, font_size, alignment, text, bold):\n",
    "    p = document.add_paragraph()\n",
    "    title = p.add_run(text)\n",
    "    title.bold = bold\n",
    "    title.font.size = Pt(font_size)\n",
    "    title.font.name = font_name\n",
    "    p.alignment = alignment\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_line_break(document):\n",
    "    p = document.add_paragraph()\n",
    "    run = p.add_run()\n",
    "    run.add_break(docx.text.run.WD_BREAK.LINE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "add_line_break(document)\n",
    "\n",
    "make_paragraph(document, 'Calibri', 14, WD_ALIGN_PARAGRAPH.CENTER, 'RESULT LINELIST FOR SARS-COV-2 WHOLE GENOME SEQUENCING (GECO-PH)', bold = True) #title\n",
    "make_paragraph(document, 'Calibri', 10.5, WD_ALIGN_PARAGRAPH.CENTER, date.today().strftime('%d %B %Y'), False)  #add date today\n",
    "\n",
    "add_line_break(document)\n",
    "#change_font_and_size_of_paragraph(title, 'Arial', 12)\n",
    "# title.bold = True\n",
    "# title.font.size = Pt(14)\n",
    "# title.font.name = 'Calibri'\n",
    "# p.alignment = WD_ALIGN_PARAGRAPH.CENTER\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # add a table to the end and create a reference variable\n",
    "# # extra row is so we can add the header row\n",
    "# header = document.add_table(rows=1, cols=1)\n",
    "# header.style = 'Table Grid'\n",
    "# header.cell(0,0).text = 'SAMPLES ENDORSED BY EPIDEMIOLOGY BUREAU'\n",
    "\n",
    "# table = document.add_table(rows=df.shape[0]+1, cols=df.shape[1])\n",
    "# table.style = 'Table Grid'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_table(df,table):\n",
    "    for j in range(df.shape[-1]):\n",
    "        table.cell(0,j).text = df.columns[j]\n",
    "    \n",
    "    for i in range(df.shape[0]):\n",
    "        for j in range(df.shape[-1]):\n",
    "            table.cell(i+1,j).text = str(df.values[i,j])\n",
    "            \n",
    "# generate_table(df,table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#add the header rows.\n",
    "# for j in range(df.shape[-1]):\n",
    "#     table.cell(0,j).text = df.columns[j]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add the rest of the data frame\n",
    "# for i in range(df.shape[0]):\n",
    "#     for j in range(df.shape[-1]):\n",
    "#         table.cell(i+1,j).text = str(df.values[i,j])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#color a cell function\n",
    "\n",
    "def set_cell_bg_color(tc, hex_color):\n",
    "    \"\"\"\n",
    "    set background shading for Header Rows\n",
    "    \"\"\"\n",
    "    tblCellProperties = tc._element.tcPr\n",
    "    clShading = OxmlElement('w:shd')\n",
    "    clShading.set(qn('w:fill'), hex_color)  # Hex of Dark Blue Shade {R:0x00, G:0x51, B:0x9E}\n",
    "    tblCellProperties.append(clShading)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set cell size\n",
    "def set_cell_width(table, column, width):\n",
    "   for cell in table.columns[column].cells:\n",
    "       cell.width = Inches(width)\n",
    "\n",
    "# set_cell_width(table, 0, 0.4)   #NO\n",
    "# set_cell_width(table, 1, 1.37)  #RITM Lab ID\n",
    "# set_cell_width(table, 2, 1.08)  #UIC\n",
    "# set_cell_width(table, 3, 0.49)  #AGE\n",
    "# set_cell_width(table, 4, 0.4)   #SEX\n",
    "# set_cell_width(table, 5, 1.1)   #PATIENT ADDRESS\n",
    "# set_cell_width(table, 6, 0.7)   #LINELIST REGION\n",
    "# set_cell_width(table, 7, 1.65)  #SENDING FACILITY\n",
    "# set_cell_width(table, 8, 0.99)  #DATA OF COLLECTION\n",
    "# set_cell_width(table, 9, 0.86)  #OUTCOME\n",
    "# set_cell_width(table, 10, 1.97) #LINEAGE\n",
    "# set_cell_width(table, 11, 1.97) #REMARKS\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# center the text in the cells\n",
    "def set_alignment(table, alignment):\n",
    "    # center the text in the cells\n",
    "    if alignment == 'center':\n",
    "        alignment = docx.enum.text.WD_ALIGN_PARAGRAPH.CENTER\n",
    "    for col in table.columns:\n",
    "        for cell in col.cells:\n",
    "            cell.paragraphs[0].alignment = alignment\n",
    "\n",
    "# set_alignment(table, 'center')\n",
    "\n",
    "def set_alignment_column(table, alignment, column_number):\n",
    "    if alignment == 'center':\n",
    "        alignment = docx.enum.text.WD_ALIGN_PARAGRAPH.CENTER\n",
    "    if alignment == 'left':\n",
    "        alignment = docx.enum.text.WD_ALIGN_PARAGRAPH.LEFT\n",
    "    for cell in table.columns[column_number].cells:\n",
    "            cell.paragraphs[0].alignment = alignment\n",
    "\n",
    "def set_font_size_column(table, column_number, size):\n",
    "    for cell in table.columns[column_number].cells:\n",
    "            paragraph = cell.paragraphs[0]\n",
    "            font = paragraph.runs[0].font\n",
    "            font.size = Pt(size)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set the font size of the text each cell\n",
    "def set_font_size(table, size):\n",
    "    for row in table.rows:\n",
    "        for cell in row.cells:\n",
    "            paragraph = cell.paragraphs[0]\n",
    "            font = paragraph.runs[0].font\n",
    "            font.size = Pt(size)\n",
    "\n",
    "# set_font_size(table, 11)\n",
    "# set_font_size(header, 12)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#make a header text bold\n",
    "def set_header_bold(table,bold):\n",
    "    #make a header text bold\n",
    "    for cell in table.rows[0].cells:\n",
    "            paragraph = cell.paragraphs[0]\n",
    "            bold_status = paragraph.runs[0]\n",
    "            bold_status.bold = bold\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "linelist_df = pd.read_csv('linelist_EB_06232022.csv')\n",
    "linelist_df.index = linelist_df.index + 1\n",
    "linelist_df.rename(columns = {'lab_id':'RITM Lab ID', 'uic':'UIC', 'age': 'AGE', 'sex': 'SEX','city_municipality': 'PATIENT ADDRESS','sending_facility':'SENDING FACILITY', 'date_specimen_collection':'DATE OF COLLECTION', 'pango_lineage': 'LINEAGE', 'linelist_region':'LINELIST REGION'}, inplace = True)\n",
    "#linelist_df = linelist_df.astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# put No. column using index\n",
    "linelist_df['NO'] = linelist_df.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#add 'COV22-' to entries in RITM LAB ID column\n",
    "def add_cov22(linelist_df): \n",
    "    #add 'COV22-' to entries in RITM LAB ID column\n",
    "    for index, row in linelist_df.iterrows():\n",
    "        linelist_df.at[index, \"RITM Lab ID\"] = 'COV22-' + row['RITM Lab ID']\n",
    "    return linelist_df\n",
    "\n",
    "linelist_df = add_cov22(linelist_df)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_outcome_column(linelist_df):\n",
    "    for index, row in linelist_df.iterrows():\n",
    "        if pd.isnull(linelist_df.at[index, \"LINEAGE\"]) == False:\n",
    "             linelist_df.at[index, \"OUTCOME\"] = 'Lineage Assigned'\n",
    "        else:\n",
    "             linelist_df.at[index, \"OUTCOME\"] = 'Lineage Unassigned'\n",
    "    return linelist_df\n",
    "\n",
    "linelist_df = make_outcome_column(linelist_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def combine_lineage(linelist_df):\n",
    "    for index, row in linelist_df.iterrows():\n",
    "        if pd.isnull(linelist_df.at[index, \"LINEAGE\"]) == False:                                            #processes lineage if value is not NAN\n",
    "             linelist_df.at[index, \"LINEAGE\"] = row['LINEAGE'] + ' (' + row['lineage'] + ')'\n",
    "        else:\n",
    "            linelist_df.at[index, \"LINEAGE\"] = ''\n",
    "    return linelist_df\n",
    "\n",
    "linelist_df = combine_lineage(linelist_df)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def set_cell_height(table):\n",
    "    row_num = 0\n",
    "    for row in table.rows:\n",
    "        if row_num == 0:\n",
    "            pass\n",
    "        else:\n",
    "            row.height = Inches(0.71)\n",
    "        row_num += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#add remarks column to dataframe\n",
    "linelist_df['REMARKS'] = np.nan\n",
    "linelist_df['REMARKS'] = ''\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#sorting and manipulating date_receieved column to become datetime objects\n",
    "date_list = linelist_df.date_received\n",
    "\n",
    "\n",
    "date_list = [datetime.strptime(datetime_str, '%m/%d/%Y') for datetime_str in date_list] #Y-M-D for date_recieved\n",
    "\n",
    "date_list = [datetime.strftime(datetime_str, '%B %d %Y') for datetime_str in date_list]\n",
    "\n",
    "\n",
    "linelist_df['date_received'] = date_list\n",
    "linelist_df['date_received'] =  pd.to_datetime(linelist_df['date_received'], format='%B %d %Y')  #makes column into datetime.datetime\n",
    "\n",
    "\n",
    "date_list = list(set(date_list))            # gets all unique dates and stores them in list\n",
    "\n",
    "\n",
    "date_list.sort(key = lambda date: datetime.strptime(date, '%B %d %Y'))      #sorts all dates\n",
    "\n",
    "\n",
    "date_list = [date_list[x:x+2] for x in range(0, len(date_list),2)]          #subset list of dates for every two date entries\n",
    "for dates in date_list:\n",
    "    dates.sort(key = lambda date: datetime.strptime(date, '%B %d %Y'))\n",
    "\n",
    "#manipulating date of collection column to make them datetime objects\n",
    "date_collected = linelist_df['DATE OF COLLECTION']\n",
    "\n",
    "date_collected = [datetime.strptime(datetime_str, '%m/%d/%Y') for datetime_str in date_collected]   #M/D/Y for date_collected\n",
    "\n",
    "date_collected = [datetime.strftime(datetime_str, '%B %d %Y') for datetime_str in date_collected]\n",
    "\n",
    "linelist_df['DATE OF COLLECTION'] = date_collected\n",
    "linelist_df['DATE OF COLLECTION'] =  pd.to_datetime(linelist_df['DATE OF COLLECTION'], format='%B %d %Y')\n",
    "\n",
    "print(linelist_df['DATE OF COLLECTION'])\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#initialize appendix df \n",
    "# 'Date of Sequencing results release', 'Sample source', 'Period of specimen collection', 'N Received', 'Rejected', 'Ongoing Investigation', 'Lineage assigned', 'Poor genome coverage', 'Key points'\n",
    "appendix_df = pd.DataFrame(columns=['Date specimens received at RITM'])\n",
    "\n",
    "# appendix_index = 0\n",
    "# for dates in date_list:\n",
    "#     if len(dates) == 2:\n",
    "#         date1 = dates[0]\n",
    "#         date2 = dates[1]\n",
    "#         appendix_df.at[appendix_index, 'Date specimens received at RITM'] = f'{date1} - {date2}'\n",
    "#         appendix_index += 1\n",
    "#     else:\n",
    "#         date1 = dates[0]\n",
    "#         appendix_df.at[appendix_index, 'Date specimens received at RITM'] = f'{date1}'\n",
    "#         appendix_index += 1\n",
    "\n",
    "date_received = []\n",
    "collection_period = []\n",
    "collection_period_subset = []\n",
    "num_samples_received = []\n",
    "rejected_samples = []\n",
    "num_rejected_samples = []\n",
    "num_lineage_assigned = []\n",
    "num_poor_genome = []\n",
    "lineage_list = []\n",
    "unique_lineage_list = []\n",
    "key_points = []\n",
    "\n",
    "#will update with date from dataframe_subsets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#subset the dataframe every two dates\n",
    "for dates in date_list:\n",
    "    \n",
    "    header = document.add_table(rows=1, cols=1)\n",
    "    header.style = 'Table Grid'\n",
    "\n",
    "    if len(dates) == 2:\n",
    "        date1 = dates[0]\n",
    "        date1 = datetime.strptime(date1, '%B %d %Y')\n",
    "        date2 = dates[1]\n",
    "        date2 = datetime.strptime(date2, '%B %d %Y')\n",
    "        print(date1)\n",
    "        print(date2)\n",
    "        linelist_subset = linelist_df[(linelist_df['date_received'] >= date1) & (linelist_df['date_received'] <= date2)]\n",
    "        date1_str = date1.strftime(\"%B %d %Y\")\n",
    "        date2_str = date2.strftime(\"%B %d %Y\")\n",
    "        date_received.append(f'{date1_str} - {date2_str}')\n",
    "                                                                                           #appends list to get date received range\n",
    "\n",
    "        collection_period_subset = list(set(linelist_subset['DATE OF COLLECTION']))\n",
    "        min_collect = min(collection_period_subset)\n",
    "        min_collect = min_collect.strftime(\"%B %d %Y\")\n",
    "        max_collect = max(collection_period_subset)\n",
    "        max_collect = max_collect.strftime(\"%B %d %Y\")\n",
    "\n",
    "        print(min_collect)\n",
    "        print(max_collect)\n",
    "        collection_period_subset = [date_str.strftime(\"%B %d %Y\") for date_str in collection_period_subset]\n",
    "\n",
    "                                                                                                                            #convert dates in dates collected to readable format\n",
    "        linelist_subset['DATE OF COLLECTION'] = linelist_subset['DATE OF COLLECTION'].dt.strftime('%m/%d/%Y')\n",
    "        print(linelist_subset['DATE OF COLLECTION'])\n",
    "\n",
    "        if len(collection_period_subset) != 1:\n",
    "            collection_period.append(min_collect + ' - ' + max_collect)                    #Appends list to get collection date range for this subset\n",
    "        else:\n",
    "            collection_period.append(min_collect)\n",
    "\n",
    "        num_samples_received.append(linelist_subset.shape[0])\n",
    "\n",
    "        rejected_samples = linelist_subset[linelist_subset['genome_coverage'].isnull()]                                     \n",
    "        num_rejected_samples.append(rejected_samples.shape[0])                                                               # number of samples where there is no genome coverage\n",
    "\n",
    "        lineage_assigned = linelist_subset[linelist_subset['OUTCOME'] == 'Lineage Assigned']\n",
    "        num_lineage_assigned.append(lineage_assigned.shape[0])                                                               #number of samples where lineage was assigned\n",
    "\n",
    "        poor_genome = linelist_subset[linelist_subset['genome_coverage'] < 70]\n",
    "        num_poor_genome.append(poor_genome.shape[0])                                                                         # number of samples where genome coverage is less than 70\n",
    "\n",
    "        lineage_list = list(linelist_subset['LINEAGE'][linelist_subset['LINEAGE'] != ''])\n",
    "        unique_lineage_list = list(set(lineage_list))\n",
    "        num_lineage_dict = {i:lineage_list.count(i) for i in unique_lineage_list}                                            # make a dictionary containing the unique lineages and their number of occurance\n",
    "\n",
    "        key_point_string = f'· {round(((lineage_assigned.shape[0]/linelist_subset.shape[0])*100),2)}% ({lineage_assigned.shape[0]}/{linelist_subset.shape[0]}) successfully assigned with lineage \\n'\n",
    "\n",
    "        for key in num_lineage_dict.keys():\n",
    "            key_point_string = key_point_string + f'· {round(((num_lineage_dict[key]/lineage_assigned.shape[0])*100),2)}% ({num_lineage_dict[key]}/{lineage_assigned.shape[0]}) {key} \\n'\n",
    "\n",
    "\n",
    "        header.cell(0,0).text = f'SAMPLES ENDORSED BY EPIDEMIOLOGY BUREAU (SAMPLES RECEIVED - {date1_str} TO {date2_str})'\n",
    "    else:\n",
    "        date1 = dates[0]\n",
    "        linelist_subset = linelist_df[(linelist_df['date_received'] >= date1)]\n",
    "        \n",
    "        date_received.append(f'{date1}')\n",
    "                                                                                                     #appends list to get date received range\n",
    "\n",
    "        collection_period_subset = list(set(linelist_subset['DATE OF COLLECTION']))\n",
    "        min_collect = min(collection_period_subset)\n",
    "        min_collect = min_collect.strftime(\"%B %d %Y\")\n",
    "        max_collect = max(collection_period_subset)\n",
    "        max_collect = max_collect.strftime(\"%B %d %Y\")\n",
    "\n",
    "        print(min_collect)\n",
    "        print(max_collect)\n",
    "        collection_period_subset = [date_str.strftime(\"%B %d %Y\") for date_str in collection_period_subset]\n",
    "        if len(collection_period_subset) != 1:\n",
    "            collection_period.append(min_collect + ' - ' + max_collect)                    #Appends list to get collection date range for this subset\n",
    "        else:\n",
    "            collection_period.append(min_collect)\n",
    "\n",
    "        linelist_subset['DATE OF COLLECTION'] = linelist_subset['DATE OF COLLECTION'].dt.strftime('%m/%d/%Y')               #convert dates in dates collected to readable format\n",
    "\n",
    "        num_samples_received.append(linelist_subset.shape[0])                                                               #number of samples in this receiving period\n",
    "\n",
    "        rejected_samples = linelist_subset[linelist_subset['genome_coverage'].isnull()]\n",
    "        num_rejected_samples.append(rejected_samples.shape[0])                                                              # number of samples where there is no genome coverage\n",
    "\n",
    "        lineage_assigned = linelist_subset[linelist_subset['OUTCOME'] == 'Lineage Assigned']\n",
    "        num_lineage_assigned.append(lineage_assigned.shape[0])                                                               #number of samples where lineage was assigned\n",
    "\n",
    "        poor_genome = linelist_subset[linelist_subset['genome_coverage'] < 70]\n",
    "        num_poor_genome.append(poor_genome.shape[0])                                                                       # number of samples where genome coverage is less than 70\n",
    "\n",
    "        lineage_list = list(linelist_subset['LINEAGE'][linelist_subset['LINEAGE'] != ''])\n",
    "        unique_lineage_list = list(set(lineage_list))\n",
    "        num_lineage_dict = {i:lineage_list.count(i) for i in unique_lineage_list}                                           # make a dictionary containing the unique lineages and their number of occurance\n",
    "    \n",
    "        key_point_string = f'· {round(((lineage_assigned.shape[0]/linelist_subset.shape[0])*100),2)}% ({lineage_assigned.shape[0]}/{linelist_subset.shape[0]}) successfully assigned with lineage \\n'\n",
    "\n",
    "        for key in num_lineage_dict.keys():\n",
    "\n",
    "            key_point_string = key_point_string + f'· {round(((num_lineage_dict[key]/lineage_assigned.shape[0])*100),2)}% ({num_lineage_dict[key]}/{lineage_assigned.shape[0]}) {key} \\n'\n",
    "\n",
    "        \n",
    "        \n",
    "        header.cell(0,0).text = f'SAMPLES ENDORSED BY EPIDEMIOLOGY BUREAU (SAMPLES RECEIVED - {date1})'\n",
    "    \n",
    "    \n",
    "    linelist_subset = linelist_subset[['NO', 'RITM Lab ID', 'UIC', 'AGE', 'SEX', 'PATIENT ADDRESS', 'LINELIST REGION', 'SENDING FACILITY', 'DATE OF COLLECTION', 'OUTCOME', 'LINEAGE','REMARKS']]\n",
    "    linelist_subset['NO'] = np.arange(len(linelist_subset)) + 1\n",
    "\n",
    "    \n",
    "\n",
    "\n",
    "    set_cell_bg_color(header.cell(0,0), '084C22')\n",
    "    set_font_size(header, 12)\n",
    "    set_header_bold(header, True)\n",
    "    \n",
    "    eb_table = document.add_table(rows=linelist_subset.shape[0]+1, cols=linelist_subset.shape[1])\n",
    "    eb_table.style = 'Table Grid'\n",
    "    generate_table(linelist_subset, eb_table)\n",
    "\n",
    "    set_cell_width(eb_table, 0, 0.4)   #NO\n",
    "    set_cell_width(eb_table, 1, 1.37)  #RITM Lab ID\n",
    "    set_cell_width(eb_table, 2, 1.08)  #UIC\n",
    "    set_cell_width(eb_table, 3, 0.49)  #AGE\n",
    "    set_cell_width(eb_table, 4, 0.4)   #SEX\n",
    "    set_cell_width(eb_table, 5, 1.1)   #PATIENT ADDRESS\n",
    "    set_cell_width(eb_table, 6, 0.7)   #LINELIST REGION\n",
    "    set_cell_width(eb_table, 7, 1.65)  #SENDING FACILITY\n",
    "    set_cell_width(eb_table, 8, 0.99)  #DATA OF COLLECTION\n",
    "    set_cell_width(eb_table, 9, 0.86)  #OUTCOME\n",
    "    set_cell_width(eb_table, 10, 1.97) #LINEAGE\n",
    "    set_cell_width(eb_table, 11, 1.97) #REMARKS\n",
    "\n",
    "    set_alignment(eb_table, 'center')\n",
    "\n",
    "    set_font_size(eb_table, 11)\n",
    "    \n",
    "    set_header_bold(eb_table, True)\n",
    "\n",
    "    set_cell_height(eb_table)\n",
    "    \n",
    "    document.add_page_break()\n",
    "\n",
    "  \n",
    "    key_points.append(key_point_string)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "appendix_df['Date specimens received at RITM'] = date_received\n",
    "appendix_df['Date of Sequencing results release'] = ''\n",
    "appendix_df['Sample source'] = 'Targeted samples (endorsed by EB)'\n",
    "appendix_df['Period of specimen collection'] = collection_period\n",
    "appendix_df['N Received'] = num_samples_received\n",
    "appendix_df['Rejected'] = num_rejected_samples\n",
    "appendix_df['Ongoing Investigation'] = ''\n",
    "appendix_df['Lineage assigned'] = num_lineage_assigned\n",
    "appendix_df['Poor genome coverage'] = num_poor_genome\n",
    "appendix_df['Key points'] = key_points"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "document.add_page_break()\n",
    "make_paragraph(document, 'Calibri', 10, WD_ALIGN_PARAGRAPH.CENTER, 'APPENDIX: RUN SUMMARY', bold = True)\n",
    "add_line_break(document)\n",
    "make_paragraph(document, 'Calibri', 10, WD_ALIGN_PARAGRAPH.LEFT, 'Table 1. Summary of sample processing details and key points. ', bold = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate table in for appendix_df in docx file \n",
    "\n",
    "table1 = document.add_table(rows=appendix_df.shape[0]+1, cols=appendix_df.shape[1])\n",
    "table1.style = 'Table Grid'\n",
    "\n",
    "generate_table(appendix_df, table1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# center text in cells\n",
    "set_alignment(table1, 'center')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#color the header row for table 1\n",
    "def color_header(table, hex_color):\n",
    "    #color the header row for table 1\n",
    "    cell_count = 0\n",
    "    for cell in table.rows[0].cells:\n",
    "        set_cell_bg_color(table.cell(0,cell_count), hex_color)\n",
    "        \n",
    "        cell_count +=1\n",
    "\n",
    "color_header(table1, '084C22')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#set the cell width and font size for table 1\n",
    "set_cell_width(table1, 0, 1.5)   #Date specimens received at RITM\n",
    "set_cell_width(table1, 1, 1.42)  #Date of Sequencing results release\n",
    "set_cell_width(table1, 2, 1.72)  #Sample source\n",
    "set_cell_width(table1, 3, 1.34)  #Period of specimen collection\n",
    "set_cell_width(table1, 4, 0.62)   #N Received\n",
    "set_cell_width(table1, 5, 0.65)   #Rejected\n",
    "set_cell_width(table1, 6, 0.82)   #Ongoing Investigation\n",
    "set_cell_width(table1, 7, 0.84)  #Lineage assigned\n",
    "set_cell_width(table1, 8, 0.82)  #Poor genome coverage\n",
    "set_cell_width(table1, 9, 3.26)  #Key points\n",
    "\n",
    "set_alignment(table1, 'center')\n",
    "\n",
    "set_font_size(table1, 9)\n",
    "\n",
    "set_alignment_column(table1, 'left', 9)\n",
    "set_font_size_column(table1, column_number = 8, size = 8)\n",
    "set_header_bold(table1, bold = True)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TABLE 2\n",
    "document.add_page_break()\n",
    "make_paragraph(document, 'Calibri', 10, WD_ALIGN_PARAGRAPH.LEFT, 'Table 2. Breakdown of results for EB-endorsed samples processed per region/province.', bold = False)\n",
    "\n",
    "breakdown_df = pd.DataFrame(columns=['REPORTNG REGION'])\n",
    "#breakdown_df['REPORTNG REGION'] = list(set(linelist_df['LINELIST REGION'].dropna()))                                 #assign all unique regions as the value of the column\n",
    "breakdown_df['REGION-PROVINCE'] = list(set(linelist_df['PATIENT ADDRESS'].dropna()))\n",
    "\n",
    "province_list = list(breakdown_df['REGION-PROVINCE'])\n",
    "\n",
    "\n",
    "classification_list = ['Cluster', 'Admitted', 'Reinfection', 'ROF', 'Unknown Exposure']\n",
    "# breakdown_df['CLASSIFICATION'] = []\n",
    "\n",
    "# test_df = pd.DataFrame(columns=['CLASSIFICATION'])\n",
    "\n",
    "lineage_list = list(set(linelist_df['LINEAGE'][linelist_df['LINEAGE'] != '']))\n",
    "test_df = pd.DataFrame(columns=['REPORTNG REGION','REGION-PROVINCE','NO. OF SAMPLES','CLASSIFICATION'] + lineage_list + ['LINEAGE NOT ASSIGNED'])\n",
    "\n",
    "classification_table = []\n",
    "\n",
    "for place in province_list:\n",
    "    for x in classification_list:\n",
    "        classification_table.append(x)\n",
    "\n",
    "\n",
    "\n",
    "test_df['CLASSIFICATION'] = classification_table\n",
    "\n",
    "test_df['NO. OF SAMPLES'] = '0'\n",
    "\n",
    "test_df[lineage_list] = '0'\n",
    "\n",
    "test_df['LINEAGE NOT ASSIGNED'] = '0'\n",
    "\n",
    "\n",
    "    \n",
    "\n",
    "# for index, row in breakdown_df.iterrows():\n",
    "#     breakdown_df.at[index, 'CLASSIFICATION'] = classification_list\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "print(lineage_list)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#initialize list\n",
    "province_data = pd.DataFrame(columns = ['linelist_region', 'province', 'classification', 'lineage'])\n",
    "\n",
    "linelist_df['LINELIST REGION'] = linelist_df['LINELIST REGION'].astype(str)\n",
    "\n",
    "#get unique list of linelist_region\n",
    "linelist_region = list(set(linelist_df['LINELIST REGION'].dropna().str.upper()))\n",
    "# make an region order \n",
    "region_order = ['CAR','1', '2', '3', '4A', '4B', '5', '6', '7', '8', '9', '10', '11', '12', '13', 'BARMM', 'NCR']\n",
    "order = {key: i for i, key in enumerate(region_order)}\n",
    "\n",
    "#sort the linelist_region\n",
    "linelist_region.sort(key=lambda val: order[val])\n",
    "\n",
    "\n",
    "str_linelist_df = linelist_df.astype(str)\n",
    "\n",
    "#subset data drame using linelist_region\n",
    "for region in linelist_region:\n",
    "    linelist_region_subset = str_linelist_df[str_linelist_df['LINELIST REGION'].str.upper() == region]\n",
    "    province_list = list(set(linelist_region_subset['PATIENT ADDRESS'].dropna().str.upper()))\n",
    "    region_str = region\n",
    "\n",
    "    #subset dataframe using province\n",
    "    for province in province_list:\n",
    "        province_subset = linelist_region_subset[linelist_region_subset['PATIENT ADDRESS'].str.upper() == province]\n",
    "        province_classification = list(province_subset['criteria_for_selection'])\n",
    "        province_lineage = list(province_subset['LINEAGE'])\n",
    "        province_classification = [x.upper() for x in province_classification]\n",
    "        province_lineage = [x.upper() for x in province_lineage]\n",
    "        province_str = province\n",
    "\n",
    "        province_data_subset = pd.DataFrame(columns = ['linelist_region', 'province', 'classification', 'lineage'])\n",
    "        province_data_subset['province'] = [province_str]\n",
    "        province_data_subset['linelist_region'] = [region_str]\n",
    "        province_data_subset['classification'] = [province_classification]\n",
    "        province_data_subset['lineage'] = [province_lineage]\n",
    "\n",
    "        province_data = pd.concat([province_data, province_data_subset])\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "       \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "lineage_list = [x.upper() for x in lineage_list]\n",
    "\n",
    "table2_df = pd.DataFrame(columns=['REPORTNG REGION','REGION-PROVINCE','NO. OF SAMPLES','CLASSIFICATION'] + lineage_list + ['LINEAGE NOT ASSIGNED'])\n",
    "\n",
    "temp_df = pd.DataFrame(columns=['REPORTNG REGION','REGION-PROVINCE','NO. OF SAMPLES','CLASSIFICATION'] + lineage_list + ['LINEAGE NOT ASSIGNED']) \n",
    "\n",
    "classification_list = [x.upper() for x in classification_list]\n",
    "\n",
    "criteria_dict = {'CLUSTER': 0, 'ADMITTED': 1, 'REINFECTION': 2, 'ROF': 3, 'UNKNOWN EXPOSURE': 4, 'ADMITTED*': 1}\n",
    "\n",
    "\n",
    "for index, row in province_data.iterrows():\n",
    "    temp_df['CLASSIFICATION'] = classification_list\n",
    "    temp_df['REPORTNG REGION'] = row['linelist_region']\n",
    "    temp_df['NO. OF SAMPLES'] = 0\n",
    "    temp_df[lineage_list] = 0\n",
    "    temp_df['LINEAGE NOT ASSIGNED'] = 0\n",
    "    temp_df['REGION-PROVINCE'] = row['province']\n",
    "    \n",
    "    for i in classification_list:\n",
    "        for j in row['classification']:\n",
    "            if (j == i) or (i in j):\n",
    "                temp_df.at[criteria_dict[i], 'NO. OF SAMPLES'] += 1\n",
    "\n",
    "    for x in row['lineage']:\n",
    "        # classification_lineage_index = row['classification'][row['lineage'].index(x)]\n",
    "        if (x == ''):\n",
    "                classification_lineage_index = row['classification'][row['lineage'].index(x)]\n",
    "                # print(classification_lineage_index)\n",
    "                temp_df.at[criteria_dict[classification_lineage_index], 'LINEAGE NOT ASSIGNED'] += 1\n",
    "\n",
    "    lineage_counter = 0\n",
    "\n",
    "    for x in row['lineage']:\n",
    "        if x in lineage_list:\n",
    "            classification_lineage_index = row['classification'][lineage_counter]\n",
    "            temp_df.at[criteria_dict[classification_lineage_index], x] += 1\n",
    "          \n",
    "        lineage_counter += 1\n",
    "\n",
    "    table2_df = pd.concat([table2_df, temp_df])\n",
    "\n",
    "print(table2_df)\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "table2_df = table2_df.reset_index(drop = True)\n",
    "\n",
    "def mark_classification_with_all_zero(table2_df):\n",
    "    for index, row in table2_df.iterrows():\n",
    "        if row['NO. OF SAMPLES'] == 0:\n",
    "            table2_df.at[index, 'CLASSIFICATION'] = row['CLASSIFICATION'] + '*'\n",
    "\n",
    "    return table2_df\n",
    "\n",
    "\n",
    "table2_df = mark_classification_with_all_zero(table2_df)\n",
    "\n",
    "\n",
    "\n",
    "#rename lineage columns add 'Lineage Assigned:'\n",
    "for x in lineage_list:\n",
    "    table2_df.rename(columns = {x: 'Lineage Assigned: ' + x}, inplace = True)\n",
    "\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "table2 = document.add_table(rows=table2_df.shape[0]+1, cols=table2_df.shape[1])\n",
    "table2.style = 'Table Grid'\n",
    "\n",
    "generate_table(table2_df, table2)\n",
    "#generate_table_with_color(table2, table2_df)\n",
    "\n",
    "# for i in range(breakdown_df.shape[0]):\n",
    "#     table2.cell(i+1,3).text = ''\n",
    "#     table_inside = table2.cell(i+1,3).add_table(rows = 5 , cols = 1 )\n",
    "#     table_inside.style = 'Table Grid'\n",
    "#     for x in range(len(classification_list)):\n",
    "#         table_inside.cell(x,0).text = classification_list[x]\n",
    "        \n",
    "# for i in range(breakdown_df.shape[0]):\n",
    "    \n",
    "#     table_inside = table2.cell(i+1,2).add_table(rows = 5 , cols = 1 )\n",
    "#     table_inside.style = 'Table Grid'\n",
    "#     for x in range(len(classification_list)):\n",
    "#         table_inside.cell(x,0).text = '0'\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Style Table 2\n",
    "#set the cell width and font size for table 2\n",
    "\n",
    "# HEADER\n",
    "color_header(table2, '084C22')\n",
    "set_font_size(table2, 9)\n",
    "set_header_bold(table2, True)\n",
    "\n",
    "\n",
    "# set_cell_width(table2, 0, 0.69)   #REPORTNG REGION\n",
    "# set_cell_width(table2, 1, 0.8)  #REGION-PROVINCE\n",
    "# set_cell_width(table2, 2, 0.66)  #NO. OF SAMPLES\n",
    "# set_cell_width(table2, 3, 1.07)  #CLASSIFICATION\n",
    "# set_cell_width(table2, 4, 0.62)   #N Received\n",
    "# set_cell_width(table2, 5, 0.65)   #Rejected\n",
    "# set_cell_width(table2, 6, 0.82)   #Ongoing Investigation\n",
    "# set_cell_width(table2, 7, 0.84)  #Lineage assigned\n",
    "# set_cell_width(table2, 8, 0.82)  #Poor genome coverage\n",
    "# set_cell_width(table2, 9, 3.26)  #Key points\n",
    "\n",
    "set_alignment(table2, 'center')\n",
    "\n",
    "set_font_size(table2, 9)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# changee font color in table using value\n",
    "def set_font_color_condition(table):\n",
    "    for row in table.rows:\n",
    "        for cell in row.cells:\n",
    "            if cell.text == '0':\n",
    "                paragraph = cell.paragraphs[0]\n",
    "                font = paragraph.runs[0].font\n",
    "                color = font.color\n",
    "                color.rgb = RGBColor(0x80, 0x80, 0x80)\n",
    "\n",
    "set_font_color_condition(table2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def set_classification_column_gray():\n",
    "    for cell in table2.columns[3].cells:\n",
    "        if '*' in cell.text:\n",
    "            cell.text = cell.text.replace('*', '')\n",
    "            paragraph = cell.paragraphs[0]\n",
    "            font = paragraph.runs[0].font\n",
    "            color = font.color\n",
    "            color.rgb = RGBColor(0x80, 0x80, 0x80) \n",
    "\n",
    "set_classification_column_gray()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def merge_same_text_in_column(column):\n",
    "    prev_cell = docx.table._Cell\n",
    "    cell_str = ''\n",
    "    for cell in table2.columns[column].cells:\n",
    "        cell_str = prev_cell.text\n",
    "        if cell_str == cell.text:\n",
    "            cell.merge(prev_cell).text = cell_str\n",
    "            cell.text = cell_str\n",
    "        prev_cell = cell\n",
    "        \n",
    "merge_same_text_in_column(0)\n",
    "merge_same_text_in_column(1)\n",
    "set_alignment_column(table2,'center',0)\n",
    "set_alignment_column(table2,'center',1)\n",
    "set_alignment_column(table2,'center',3)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save the doc\n",
    "document.save('test.docx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "8809d651a54ee0c85b0695e0d15f857d8e1195256997b34db384d2391572ff56"
  },
  "kernelspec": {
   "display_name": "Python 3.10.4 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
